Tesla coop steer

dzid26 opened the pull request:
Adds steer angle offset based on measured driver torque value. 
Uses +/-0.2Nm deadzone to filter out torque sensor offset and other disturbances. 

- [x] target angle from lat accel
- [ ] enable full angle range and higher steering velocity at low speed - but ensure stability
- [ ] ~~learn steering torque offset~~ eh, just increase deadzone to 0.5Nm, since it's somewhat angle dependant
- [ ] 100Hz angle request won't work?


-------
<!--
We need these details to verify your pull request.

Find your device's dongle ID and a route at https://connect.comma.ai.
Ideally, the route is recorded with the exact branch of your pull request.
-->
Validation
* Dongle ID: 
* Route:

dzid26 committed:
virtual spring for angle control

sshane commented on opendbc/car/tesla/carcontroller.py:
```diff
@@ -76,9 +79,16 @@ def update(self, CC, CS, now_nanos):
+    # ignore potential steer bias
+    steeringTorqueDeadzone = CS.out.steeringTorque - np.clip(CS.out.steeringTorque, -STEER_BIAS_MAX, STEER_BIAS_MAX)
+    # create virtual spring effect
+    steeringAngleDeg += steeringTorqueDeadzone * STEER_VIRTUAL_SPRING_COEFF
```
You can't really directly modify angle from torque, that's not how you drive your car. torque translates to lateral accel. I tried a fork that had this and you couldn't override at all at low speeds. Need to translate to lateral accel, then translate to angle to fix,

dzid26 replied:
Agree. Also this using lat accel solves oversensitivity at highway speeds.  

At low speed it will be tricky due to high gain (torque to angle) feedback loop. To prevent oscillations, probably the real driver torque will have to be decoupled from steering inertia torque.  I am thinking maybe a feedforward steering inertia estimator based on the commanded angle could help stabilize it.  

For now I clipped the gain so it doesn't oscillate. - This feature is already is a massive driving experience improvement. 

One thing that should be fixed is torque sensor offset (my steering has pretty consistent -0.2Nm ).  Would be interesting to check values across the fleet.  I don't want to increase a deadzone too much to filter this out else it may become noticable. The offset should be learnt and saved, so I guess it should be added to paramsd ?

dzid26 committed:
lateral acc
Squashed commit of the following:

commit 0738c5a3a30847f8dc918572791e62472c5ce4f4
Author: dzid26 <dzidmail@gmail.com>
Date:   Tue Jul 22 16:37:16 2025 +0100

    tune lat acc

commit 7c968422a5ef50e00a988fa2701044c08f1fe60a
Author: dzid26 <dzidmail@gmail.com>
Date:   Tue Jul 22 14:03:20 2025 +0100

    set desired lat acceleration, use simple target for now

commit 429fcf5ec6838297037d31528e3f79ce3848c371
Author: dzid26 <dzidmail@gmail.com>
Date:   Tue Jul 22 00:15:54 2025 +0100

    physical steering moment

commit a071a74cf14812c661972082f65cc8a81c1c5f12
Author: dzid26 <dzidmail@gmail.com>
Date:   Mon Jul 21 23:21:28 2025 +0100

    added damping

commit 2e33e0d53dcfe06af525adac09e481e412c853c0
Author: dzid26 <dzidmail@gmail.com>
Date:   Mon Jul 21 00:56:32 2025 +0100

    reset/clip last delta when changing direction

commit 361cf220c3f50d20453599b4f1a4cdf59aa73d35
Author: dzid26 <dzidmail@gmail.com>
Date:   Sun Jul 20 22:13:18 2025 +0100

    take outer rate limit into the account

commit 3d04dcfcc858bca763df53b8b834b6cf5db70c09
Author: dzid26 <dzidmail@gmail.com>
Date:   Sun Jul 20 19:55:27 2025 +0100

    fixes and refactor acc rate lim

commit b3b24aff85c1cd28d6e235dc5270aaaa12eb451b
Author: dzid26 <dzidmail@gmail.com>
Date:   Sun Jul 20 13:52:57 2025 +0100

    fix rate limit. only limit in torque direction

commit 096c9ec50e5ef1628abfca15c96fe62fc4bf8467
Author: dzid26 <dzidmail@gmail.com>
Date:   Sat Jul 19 18:52:41 2025 +0100

    inertia ff acceleration limit

commit a3aee14ee4ed8477ca46954fcb8baf6d3411be5d
Author: dzid26 <dzidmail@gmail.com>
Date:   Sat Jul 19 18:14:16 2025 +0100

    inertia compensated driver torque

commit 10aeec084f8f8623d6eb90340890304baa870d3e
Author: dzid26 <dzidmail@gmail.com>
Date:   Sat Jul 19 14:37:55 2025 +0100

    rate lim lat acc angle

commit 48675dd9c02b6c4e17f7ecdb1fb2ef90650bbed0
Author: dzid26 <dzidmail@gmail.com>
Date:   Sat Jul 19 00:58:05 2025 +0100

    lim max angle from lat acc

commit 78e56b01af042b6f3a8c1da55391ee41a5b2d1a5
Author: dzid26 <dzidmail@gmail.com>
Date:   Fri Jul 18 17:10:41 2025 +0100

    use angle from lat accel

commit 6f004b8985a179ddb43b9197c3394f671e2045be
Author: dzid26 <dzidmail@gmail.com>
Date:   Thu Jul 10 01:39:12 2025 +0100

    Use deadzone to deal with driver torque bias

commit dd0b70b9e23b4bd33d6e46a6620f06a15845b61b
Author: dzid26 <dzidmail@gmail.com>
Date:   Mon Jul 7 23:52:48 2025 +0100

    Log SAS offset.

commit eaa173ecaee615adafdfc0a0bb5d0f7a4caee89f
Author: dzid26 <dzidmail@gmail.com>
Date:   Mon Jul 7 23:51:54 2025 +0100

    Virtual spring when overriding

commit 207f7702eab3fe5c98116a19fa4314d3c7e87ebe
Author: dzid26 <dzidmail@gmail.com>
Date:   Mon Jul 7 19:34:46 2025 +0100

    calc eps force in terms of steering column shaft

commit d3dde51b67cfbde3efcd2c5e413cb6cee34eb319
Author: dzid26 <dzidmail@gmail.com>
Date:   Mon Jul 7 18:07:08 2025 +0100

    lighter steer press
Revert "Slimming `carState`, part 2 (#2508)"

This reverts commit 019038b3de8e3d7a2239f684061c1a31230f5c61.

dzid26 committed:
Revert "Slimming `carState`, part 2 (#2508)"

This reverts commit 019038b3de8e3d7a2239f684061c1a31230f5c61.

dzid26 commented:
Better implementation is here https://github.com/dzid26/opendbc/commit/6331d05cf26a63dccd9dd41a5ed6a894d8eb876d ([dzid26/vtb](https://github.com/dzid26/opendbc/commits/vtb/))
